src/finalizer/utils/opStack.ts

```diff
@@ -182,34 +211,230 @@ export async function opStackFinalizer(
+        },
+      },
+      l2OutputOracle: {
+        [sourceId]: {
+          address: VIEM_OP_STACK_CHAINS[chainId].contracts?.l2OutputOracle?.[sourceId]?.address ?? viem.zeroAddress,
+        },
+      },
+      disputeGameFactory: {
+        [sourceId]: {
+          address: VIEM_OP_STACK_CHAINS[chainId].contracts?.disputeGameFactory?.[sourceId]?.address ?? viem.zeroAddress,
+        },
+      },
+    },
+  };
 
-  const callData = [...proofs.callData, ...finalizations.callData];
-  const crossChainTransfers = [...proofs.withdrawals, ...finalizations.withdrawals];
+  const withdrawalStatuses: string[] = [];
+  await mapAsync(events, async (event, i) => {
+    // Useful information for event:
+    const l1TokenInfo = getL1TokenInfo(event.l2TokenAddress, chainId);
+    const amountFromWei = convertFromWei(event.amountToReturn.toString(), l1TokenInfo.decimals);
 
-  return { callData, crossChainMessages: crossChainTransfers };
+    const receipt = await publicClientL2.getTransactionReceipt({
+      hash: event.transactionHash as `0x${string}`,
+    });
+    const withdrawal = getWithdrawals(receipt)[logIndexesForMessage[i]];
+    const withdrawalStatus: GetWithdrawalStatusReturnType = await publicClientL1.getWithdrawalStatus({
+      receipt,
+      chain: undefined,
+      targetChain: viemOpStackTargetChainParam,
+      logIndex: logIndexesForMessage[i],
+    });
+    withdrawalStatuses.push(withdrawalStatus);
+    if (withdrawalStatus === "ready-to-prove") {
+      const l2Output = await publicClientL1.getL2Output({
+        chain: undefined,
+        l2BlockNumber: BigInt(event.blockNumber),
+        targetChain: viemOpStackTargetChainParam,
+      });
+      const { l2OutputIndex, outputRootProof, withdrawalProof } = await publicClientL2.buildProveWithdrawal({
+        chain: undefined,
+        withdrawal,
+        output: l2Output,
+      });
+      const proofArgs = [withdrawal, l2OutputIndex, outputRootProof, withdrawalProof];
+      const callData = await crossChainMessenger.populateTransaction.proveWithdrawalTransaction(...proofArgs);
+      viemTxns.callData.push({
+        callData: callData.data,
+        target: crossChainMessenger.address,
+      });
+      viemTxns.withdrawals.push({
+        originationChainId: chainId,
+        l1TokenSymbol: l1TokenInfo.symbol,
+        amount: amountFromWei,
+        type: "misc",
+        miscReason: "proof",
+        destinationChainId: hubPoolClient.chainId,
+      });
+    } else if (withdrawalStatus === "waiting-to-finalize") {
+      const { seconds } = await publicClientL1.getTimeToFinalize({
+        chain: undefined,
+        withdrawalHash: withdrawal.withdrawalHash,
+        targetChain: viemOpStackTargetChainParam,
+      });
+      logger.debug({
+        at: `${getNetworkName(chainId)}Finalizer`,
+        message: `Withdrawal ${event.transactionHash} for ${amountFromWei} of ${
+          l1TokenInfo.symbol
+        } is in challenge period for ${seconds / 60 / 60} hours`,
```

Comment:
Might be nice to avoid a long floating number:

```suggestion
        } is in challenge period for ${Math.floor(seconds / 3600)} hours`,
```